### PR TITLE
diktat-maven-plugin

### LINKED ISSUE #261: Maven plugin: diktat should have it's own maven plugin
Describe functionality and prepare a POC

### PR DESCRIPTION
**Pull request by petertrr**
### What's done:
* Initial implementation
* Mojos for checking and fixing
* Set diktat yml config via plugin property

This pull request closes #261

### Fixme:
* [x] extract parameter descriptions to plugin.xml
* [x] provide diktat config file as a parameter
* [x] mojo for fixing
* [ ] fail-fast option (can be done later in separate PR)
* [ ] goal for config generation (can be done later in separate PR)
* [ ] choose ktlint reporter via plugin config (can be done later in separate PR)
* [x] Update docs with new plugin usage instructions

### CODE REVIEW

File: .github/workflows/diktat_snapshot.yml [hunk 1]
```diff
@@ -34,17 +34,11 @@ jobs:
       - name: Build and install
         # install diktat with version 'base_version-pre' to local maven repo
         run: mvn -B clean install -DskipTests=true
-      - name: Revert project version and set proper version for antrun
+      - name: Revert project version and set proper version for diktat check
```
Comment (akuleshov7) [inline .github/workflows/diktat_snapshot.yml:37]: how about changing readme also? Adding it as a a separate chapter

Comment (petertrr) [reply .github/workflows/diktat_snapshot.yml:37]: Yes, will add

Comment (petertrr) [reply .github/workflows/diktat_snapshot.yml:37]: Added, please take a look

File: diktat-maven-plugin/src/main/kotlin/org/cqfn/diktat/plugin/maven/DiktatBaseMojo.kt [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                file.isDirectory || file.extension.let { it == "kt" || it == "kts" }
+            }
+            .filter { it.isFile }
+            .forEach { file ->
+                log.info("Checking file $file")
+                val text = file.readText()
+                try {
+                    reporter.before(file.path)
+                    val params =
+                        KtLint.Params(
```
Comment (akuleshov7) [inline diktat-maven-plugin/src/main/kotlin/org/cqfn/diktat/plugin/maven/DiktatBaseMojo.kt:110]: I guess ktlint already has a recursive walk through

File: diktat-maven-plugin/src/main/kotlin/org/cqfn/diktat/plugin/maven/DiktatBaseMojo.kt [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                                cb = { lintError, isCorrected ->
+                                    reporter.onLintError(file.path, lintError, isCorrected)
+                                    lintErrors.add(lintError)
+                                },
+                                debug = debug
+                        )
+                    runAction(params)
+                    reporter.after(file.path)
+                } catch (e: RuleExecutionException) {
+                    log.error("Received exception", e)
```
Comment (akuleshov7) [inline diktat-maven-plugin/src/main/kotlin/org/cqfn/diktat/plugin/maven/DiktatBaseMojo.kt:125]: please provide more information in error

File: diktat-maven-plugin/src/main/kotlin/org/cqfn/diktat/plugin/maven/DiktatMojo.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+     * @param params instance of [KtLint.Params] used in analysis
+     */
+    override fun runAction(params: KtLint.Params) {
+        val fileName = params.fileName
+        val filePath = params.userData["file_path"] ?: error("File path should be provided")
+        val fileContent = File(filePath).readText(charset("UTF-8"))
+        val formattedText = KtLint.format(params)
+        if (fileContent !== formattedText) {
+            log.info("Original and formatted content differ, writing to $fileName...")
+            File(filePath).writeText(formattedText, charset("UTF-8"))
```
Comment (akuleshov7) [inline diktat-maven-plugin/src/main/kotlin/org/cqfn/diktat/plugin/maven/DiktatMojo.kt:41]: are you sure we really need to write anything to a file? 
What's with the original writer from ktlint?

WHat you gonna do with \n. I guess you need to convert newlines

**Review by akuleshov7 — APPROVED** [review]
partially lgtm, but need some clarification

File: diktat-maven-plugin/src/main/kotlin/org/cqfn/diktat/plugin/maven/DiktatMojo.kt (hunk 4 above)
Comment (petertrr) [reply diktat-maven-plugin/src/main/kotlin/org/cqfn/diktat/plugin/maven/DiktatMojo.kt:41]: Substitution of `\n` is done inside `KtLint.lint/format`.

Calling any actions on input files is done on a higher level, inside `Main.kt` in ktlint. If we want to customize behavior, e.g. have a way to provide path to diktat config file, we should be able to change that logic, that's why I re-implemented part of it.

File: diktat-maven-plugin/src/main/kotlin/org/cqfn/diktat/plugin/maven/DiktatBaseMojo.kt (hunk 3 above)
Comment (petertrr) [reply diktat-maven-plugin/src/main/kotlin/org/cqfn/diktat/plugin/maven/DiktatBaseMojo.kt:125]: Done

File: diktat-maven-plugin/pom.xml [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+                    <execution>
+                        <id>compile</id>
+                        <phase>compile</phase>
+                        <goals>
+                            <goal>compile</goal>
+                        </goals>
+                        <configuration>
+                            <sourceDirs>
+                                <source>src/main/java</source>
+                                <source>src/main/kotlin</source>
```
[the remark this answers is not in the document]
Comment (petertrr) [reply diktat-maven-plugin/pom.xml:69]: Removed, I wasn't initially sure if the plugin can be done entirely in kotlin.

File: diktat-maven-plugin/src/main/kotlin/org/cqfn/diktat/plugin/maven/DiktatMojo.kt (hunk 4 above)
Comment (akuleshov7) [reply diktat-maven-plugin/src/main/kotlin/org/cqfn/diktat/plugin/maven/DiktatMojo.kt:41]: no, no need to customize it

Comment (petertrr) [reply diktat-maven-plugin/src/main/kotlin/org/cqfn/diktat/plugin/maven/DiktatMojo.kt:41]: I mean initial logic, including rule set construction. All file processing related stuff is done inside `KtLint.lint` and I don't override anything there.

**Pull request merged**